metrique-aggregation/src/histogram.rs

```diff
@@ -0,0 +1,439 @@
+//!
+//! ## AtomicExponentialAggregationStrategy
+//!
+//! Thread-safe version of exponential bucketing. Use with [`crate::histogram::AtomicHistogram`] when you need
+//! to record values from multiple threads concurrently:
+//!
+//! ```
+//! use metrique_aggregation::histogram::{AtomicHistogram, AtomicExponentialAggregationStrategy};
+//! use std::time::Duration;
+//!
+//! let histogram: AtomicHistogram<Duration, AtomicExponentialAggregationStrategy> =
+//!     AtomicHistogram::new(AtomicExponentialAggregationStrategy::new());
+//! ```
+//!
+//! ## SortAndMerge
+//!
+//! Stores all observations exactly and sorts them on emission:
+//!
+//! - Perfect precision - no bucketing error
+//! - Memory usage grows with observation count
+//! - Slower drain operation due to sorting
+//!
+//! Use this when you need exact values and have a bounded number of observations (typically < 1000).
+
+use metrique_core::CloseValue;
+use metrique_writer::{MetricFlags, MetricValue, Observation, Value, ValueWriter};
+use smallvec::SmallVec;
+use std::marker::PhantomData;
+
+/// Strategy for aggregating observations in a histogram.
+///
+/// Implementations determine how values are stored and converted to observations
+/// when the histogram is closed.
+pub trait AggregationStrategy {
+    /// Record a single observation.
+    fn record(&mut self, value: f64);
+
+    /// Drain all observations and return them as a vector.
+    ///
+    /// This resets the strategy's internal state.
+    fn drain(&mut self) -> Vec<Observation>;
+}
+
+/// Thread-safe strategy for aggregating observations in a histogram.
+///
+/// Like [`AggregationStrategy`] but allows recording values through a shared reference.
+pub trait AtomicAggregationStrategy {
+    /// Record a single observation through a shared reference.
+    fn record(&self, value: f64);
+
+    /// Drain all observations and return them as a vector.
+    ///
+    /// This resets the strategy's internal state.
+    fn drain(&self) -> Vec<Observation>;
+}
+
+/// A histogram that collects multiple observations and emits them as a distribution.
+///
+/// Use this when you have many observations of the same metric within a single unit of work.
+/// The histogram aggregates values in memory and emits them as a single metric entry.
+///
+/// Requires `&mut self` to add values. For thread-safe access, use [`AtomicHistogram`].
+pub struct Histogram<T, S = ExponentialAggregationStrategy> {
+    strategy: S,
+    _value: PhantomData<T>,
+}
+
+impl<T, S: AggregationStrategy> Histogram<T, S> {
+    /// Create a new histogram with the given aggregation strategy.
+    pub fn new(strategy: S) -> Self {
+        Self {
+            strategy,
+            _value: PhantomData,
+        }
+    }
+
+    /// Add a value to the histogram.
+    ///
+    /// The value is converted to observations using the metric value's implementation,
+    /// then recorded in the aggregation strategy.
+    pub fn add_value(&mut self, value: T)
```

Comment:
Is it possible to do some unit management here? (Tho we want to add both Duration and f64, which means we can't condition on MetricValue being anything in particular).

Or maybe unit management does not make sense in an "aggregated" context, and there should just be a comment (or an assertion?) that you should be consistent with units.

Reply:
fixed in https://github.com/awslabs/metrique/pull/143/commits/f383d9661d134e159086ac3d874e26135894c552

There is a separate issue to work on which is that the Durations are only going to store millisecond precision. I have some ideas to address that but I'll address that in a follow up